src/irmin/contents.ml

```diff
@@ -32,6 +32,139 @@ module Cstruct = struct
   let of_string s = Ok (Cstruct.of_string s)
 end
 
+type json = [
+  | `Null
+  | `Bool of bool
+  | `String of string
+  | `Float of float
+  | `O of (string * json) list
+  | `A of json list
+]
+
+let json =
+  let open Type in
+  mu (fun ty ->
+  variant "json" (fun null bool string float obj arr -> function
+    | `Null -> null
+    | `Bool b -> bool b
+    | `String s -> string s
+    | `Float f -> float f
+    | `O o -> obj o
+    | `A a -> arr a)
+  |~ case0 "null" `Null
+  |~ case1 "bool" bool (fun x -> `Bool x)
+  |~ case1 "string" string (fun x -> `String x)
+  |~ case1 "float" float (fun x -> `Float x)
+  |~ case1 "object" (list (pair string ty)) (fun obj -> `O obj)
+  |~ case1 "array" (list ty) (fun arr -> `A arr)
+  |> sealv)
+
+module Json = struct
+  type t = (string * json) list
+
+  let t = Type.(list (pair string json))
+
+  let merge_object a b =
```

Comment:
I think you will have to define a proper 3-way merge function for JSON objects: you need to detect if a value has been updated or if it's a real conflict.